File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationHelper.java
package com.mapbox.services.android.navigation.v5.navigation;

import android.location.Location;

import com.mapbox.services.Constants;
import com.mapbox.services.android.navigation.v5.milestone.Milestone;
import com.mapbox.services.android.navigation.v5.offroute.OffRoute;
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
import com.mapbox.services.android.navigation.v5.snap.Snap;
import com.mapbox.services.android.telemetry.utils.MathUtils;
import com.mapbox.services.api.directions.v5.models.DirectionsRoute;
import com.mapbox.services.api.directions.v5.models.LegStep;
import com.mapbox.services.api.utils.turf.TurfConstants;
import com.mapbox.services.api.utils.turf.TurfMeasurement;
import com.mapbox.services.api.utils.turf.TurfMisc;
import com.mapbox.services.commons.geojson.Feature;
import com.mapbox.services.commons.geojson.LineString;
import com.mapbox.services.commons.geojson.Point;
import com.mapbox.services.commons.models.Position;
import com.mapbox.services.commons.utils.PolylineUtils;

import java.util.ArrayList;
import java.util.List;

import static com.mapbox.services.Constants.PRECISION_6;

class NavigationHelper {

  // Always get the closest position on the route to the actual
  // raw location so that can accurately calculate values.
  static Position userSnappedToRoutePosition(Location location, int legIndex, int stepIndex,
                                             DirectionsRoute route) {
    Point locationToPoint = Point.fromCoordinates(
      new double[] {location.getLongitude(), location.getLatitude()}
    );

    // Decode the geometry
    List<Position> coords = PolylineUtils.decode(route.getLegs().get(legIndex).getSteps().get(stepIndex).getGeometry(),
      PRECISION_6);

    // Uses Turf's pointOnLine, which takes a Point and a LineString to calculate the closest
    // Point on the LineString.
    Feature feature = TurfMisc.pointOnLine(locationToPoint, coords);
    return ((Point) feature.getGeometry()).getCoordinates();
  }


  static String buildInstructionString(RouteProgress routeProgress, Milestone milestone) {
    if (milestone.getInstruction() != null) {
      // Create a new custom instruction based on the Instruction packaged with the Milestone
      return milestone.getInstruction().buildInstruction(routeProgress);
    } else {
      return "";
    }
  }

  static double getStepDistanceRemaining(Position snappedPosition, int legIndex,
                                         int stepIndex, DirectionsRoute directionsRoute) {
    List<LegStep> steps = directionsRoute.getLegs().get(legIndex).getSteps();
    Position nextManeuverPosition = getNextManeuverPosition(stepIndex, steps);
    // If the users snapped position equals the next maneuver position, the distance remaining is zero.
    if (snappedPosition.equals(nextManeuverPosition)) {
      return 0;
    }
    LineString slicedLine = TurfMisc.lineSlice(
      Point.fromCoordinates(snappedPosition),
      Point.fromCoordinates(nextManeuverPosition),
      LineString.fromPolyline(steps.get(stepIndex).getGeometry(), PRECISION_6)
    );
    return TurfMeasurement.lineDistance(slicedLine, TurfConstants.UNIT_METERS);
  }

  static double getLegDistanceRemaining(double stepDistanceRemaining, int legIndex,
                                        int stepIndex, DirectionsRoute directionsRoute) {
    List<LegStep> steps = directionsRoute.getLegs().get(legIndex).getSteps();
    if ((steps.size() < stepIndex + 1)) {
      return stepDistanceRemaining;
    }
    for (int i = stepIndex + 1; i < steps.size(); i++) {
      stepDistanceRemaining += steps.get(i).getDistance();
    }
    return stepDistanceRemaining;
  }

  static double getRouteDistanceRemaining(double legDistanceRemaining, int legIndex,
                                          DirectionsRoute directionsRoute) {
    if (directionsRoute.getLegs().size() < 2) {
      return legDistanceRemaining;
    }

    for (int i = legIndex + 1; i < directionsRoute.getLegs().size(); i++) {
      legDistanceRemaining += directionsRoute.getLegs().get(i).getDistance();
    }
    return legDistanceRemaining;
  }

  static Position getNextManeuverPosition(int stepIndex, List<LegStep> steps) {
    // If there is an upcoming step, use it's maneuver as the position.
    if (steps.size() > (stepIndex + 1)) {
      return steps.get(stepIndex + 1).getManeuver().asPosition();
    }
    // TODO does the last step in directions response always a single point representing the maneuver?
    // Decode the geometry
    List<Position> coords
      = PolylineUtils.decode(steps.get(stepIndex).getGeometry(), Constants.PRECISION_6);
    return coords.size() > 1 ? coords.get(coords.size() - 1) : coords.get(coords.size());
  }

  /**
   * Checks whether the user's bearing matches the next step's maneuver provided bearingAfter variable. This is one of
   * the criteria's required for the user location to be recognized as being on the next step or potentially arriving.
   *
   * @param userLocation  the location of the user
   * @param routeProgress used for getting route information
   * @return boolean true if the user location matches (using a tolerance) the final heading
   * @since 0.2.0
   */
  static boolean bearingMatchesManeuverFinalHeading(Location userLocation, RouteProgress routeProgress,
                                                    double maxTurnCompletionOffset) {
    if (routeProgress.currentLegProgress().upComingStep() == null) {
      return false;
    }

    // Bearings need to be normalized so when the bearingAfter is 359 and the user heading is 1, we count this as
    // within the MAXIMUM_ALLOWED_DEGREE_OFFSET_FOR_TURN_COMPLETION.
    double finalHeading = routeProgress.currentLegProgress().upComingStep().getManeuver().getBearingAfter();
    double finalHeadingNormalized = MathUtils.wrap(finalHeading, 0, 360);
    double userHeadingNormalized = MathUtils.wrap(userLocation.getBearing(), 0, 360);
    return MathUtils.differenceBetweenAngles(finalHeadingNormalized, userHeadingNormalized)
      <= maxTurnCompletionOffset;
  }

  /**
   * This is used when a user has completed a step maneuver and the indices need to be incremented.
   * The main purpose of this class is to determine if an additional leg exist and the step index
   * has met the first legs total size, a leg index needs to occur and step index should be reset.
   * Otherwise, the step index is incremented while the leg index remains the same.
   * <p>
   * Rather than returning an int array, a new instance of Navigation Indices gets returned. This
   * provides type safety and making the code a bit more readable.
   * </p>
   *
   * @param routeProgress   need a routeProgress in order to get the directions route leg list size
   * @param previousIndices used for adjusting the indices
   * @return a {@link NavigationIndices} object which contains the new leg and step indices
   */
  static NavigationIndices increaseIndex(RouteProgress routeProgress, NavigationIndices previousIndices) {
    // Check if we are in the last step in the current routeLeg and iterate it if needed.
    if (previousIndices.stepIndex() >= routeProgress.directionsRoute().getLegs().get(routeProgress.legIndex()).getSteps().size() - 2
      && previousIndices.legIndex() < routeProgress.directionsRoute().getLegs().size() - 1) {
      return NavigationIndices.create((previousIndices.legIndex() + 1), 0);
    }
    return NavigationIndices.create(previousIndices.legIndex(), (previousIndices.stepIndex() + 1));
  }

  // TODO needs work
  static List<Milestone> checkMilestones(RouteProgress previousRouteProgress,
                                         RouteProgress routeProgress, MapboxNavigation mapboxNavigation) {
    List<Milestone> milestones = new ArrayList<>();
    for (Milestone milestone : mapboxNavigation.getMilestones()) {
      if (milestone.isOccurring(previousRouteProgress, routeProgress)) {
        milestones.add(milestone);
      }
    }
    return milestones;
  }

  static boolean isUserOffRoute(NewLocationModel newLocationModel, RouteProgress routeProgress) {
    OffRoute offRoute = newLocationModel.mapboxNavigation().getOffRouteEngine();
    if (offRoute == null) {
      return false;
    }
    return offRoute.isUserOffRoute(newLocationModel.location(), routeProgress,
      newLocationModel.mapboxNavigation().options());
  }

  static Location getSnappedLocation(MapboxNavigation mapboxNavigation, Location location, RouteProgress routeProgress) {
    Snap snap = mapboxNavigation.getSnapEngine();
    return snap.getSnappedLocation(location, routeProgress);
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationHelper.kt
package com.mapbox.services.android.navigation.v5.internal.navigation

import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.api.directions.v5.models.LegAnnotation
import com.mapbox.api.directions.v5.models.RouteLeg
import com.mapbox.core.constants.Constants.PRECISION_6
import com.mapbox.geojson.Point
import com.mapbox.geojson.utils.PolylineUtils
import com.mapbox.services.android.navigation.v5.milestone.Milestone
import com.mapbox.services.android.navigation.v5.routeprogress.CurrentLegAnnotation
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull
import java.util.ArrayList

/**
 * This contains several single purpose methods that help out when a new location update occurs and
 * calculations need to be performed on it.
 */
internal object NavigationHelper {

    private const val INDEX_ZERO = 0
    private const val EMPTY_STRING = ""

    /**
     * When a milestones triggered, it's instruction needs to be built either using the provided
     * string or an empty string.
     */
    fun buildInstructionString(
        routeProgress: RouteProgress,
        milestone: Milestone
    ): String =
        milestone.instruction?.buildInstruction(routeProgress) ?: EMPTY_STRING

    /**
     * Takes in the leg distance remaining value already calculated and if additional legs need to be
     * traversed along after the current one, adds those distances and returns the new distance.
     * Otherwise, if the route only contains one leg or the users on the last leg, this value will
     * equal the leg distance remaining.
     */
    @JvmStatic
    fun routeDistanceRemaining(
        legDistanceRemaining: Double,
        legIndex: Int,
        directionsRoute: DirectionsRoute
    ): Double {
        var distanceRemaining = legDistanceRemaining
        directionsRoute.legs()?.let { legs ->
            val legsSize = legs.size
            if (legsSize < 2) {
                return distanceRemaining
            }

            for (i in legIndex + 1 until legsSize) {
                distanceRemaining += legs[i].distance() ?: 0.0
            }
        }
        return distanceRemaining
    }

    /**
     * Given the current [DirectionsRoute] and leg / step index,
     * return a list of [Point] representing the current step.
     *
     * This method is only used on a per-step basis as [PolylineUtils.decode]
     * can be a heavy operation based on the length of the step.
     *
     * Returns null if index is invalid.
     *
     * @param directionsRoute for list of steps
     * @param legIndex to get current step list
     * @param stepIndex to get current step
     * @return list of [Point] representing the current step
     */

    @JvmStatic
    fun decodeStepPoints(
        directionsRoute: DirectionsRoute,
        currentPoints: List<Point>?,
        legIndex: Int,
        stepIndex: Int
    ): List<Point>? {
        val legs = directionsRoute.legs()
        legs?.let { legs ->
            if (legs.isEmpty()) return currentPoints
            val steps = legs[legIndex].steps()
            steps?.let { steps ->
                if (steps.isEmpty()) return currentPoints
                val invalidStepIndex = stepIndex < 0 || stepIndex > steps.size - 1
                if (invalidStepIndex) return currentPoints
                val step = steps[stepIndex]
                step?.let { step ->
                    val stepGeometry = step.geometry()
                    stepGeometry?.let {
                        return PolylineUtils.decode(stepGeometry, PRECISION_6)
                    }
                }
            }
        }
        return currentPoints
    }

    /**
     * Given a list of distance annotations, find the current annotation index.  This index retrieves the
     * current annotation from any provided annotation list in [LegAnnotation].
     *
     * @param currentLegAnnotation current annotation being traveled along
     * @param leg holding each list of annotations
     * @param legDistanceRemaining to determine the new set of annotations
     * @return a current set of annotation data for the user's position along the route
     */
    @JvmStatic
    fun createCurrentAnnotation(
        currentLegAnnotation: CurrentLegAnnotation?,
        leg: RouteLeg,
        legDistanceRemaining: Double
    ): CurrentLegAnnotation? {
        leg.annotation()?.distance()?.let { distanceList ->
            if (distanceList.isEmpty()) return null

            val annotationBuilder = CurrentLegAnnotation.builder()
            val annotationIndex = findAnnotationIndex(
                currentLegAnnotation,
                annotationBuilder,
                leg,
                legDistanceRemaining,
                distanceList
            )
            annotationBuilder.distance(distanceList[annotationIndex])
            val durationList = leg.annotation()?.duration()
            durationList?.let {
                annotationBuilder.duration(durationList[annotationIndex])
            }
            val speedList = leg.annotation()?.speed()
            speedList?.let {
                annotationBuilder.speed(speedList[annotationIndex])
            }
            val maxSpeedList = leg.annotation()?.maxspeed()
            maxSpeedList?.let {
                annotationBuilder.maxspeed(maxSpeedList[annotationIndex])
            }
            val congestionList = leg.annotation()?.congestion()
            congestionList?.let {
                annotationBuilder.congestion(congestionList[annotationIndex])
            }
            annotationBuilder.index(annotationIndex)
            return annotationBuilder.build()
        } ?: return null
    }

    private fun findAnnotationIndex(
        currentLegAnnotation: CurrentLegAnnotation?,
        annotationBuilder: CurrentLegAnnotation.Companion.Builder,
        leg: RouteLeg,
        legDistanceRemaining: Double,
        distanceAnnotationList: List<Double?>
    ): Int {
        val legDistances = ArrayList(distanceAnnotationList)
        val totalLegDistance = leg.distance()
        totalLegDistance?.let {
            val distanceTraveled = totalLegDistance - legDistanceRemaining
            var distanceIndex = 0
            var annotationDistancesTraveled = 0.0
            ifNonNull(currentLegAnnotation) {
                distanceIndex = it.index
                annotationDistancesTraveled = it.distanceToAnnotation
            }
            for (i in distanceIndex until legDistances.size) {
                val distance = legDistances[i]
                distance?.let {
                    annotationDistancesTraveled += it
                    if (annotationDistancesTraveled > distanceTraveled) {
                        val distanceToAnnotation = annotationDistancesTraveled - it
                        annotationBuilder.distanceToAnnotation(distanceToAnnotation)
                        return i
                    }
                }
            }
        }
        return INDEX_ZERO
    }
}
